File: PartitionStr.cpp
//Array Partition for Strings in C++
#include <cstdlib>
#include <string.h>
#include<iostream>

void quickSortMain(char items[][10], int count);
void quickSort(char items[][10], int left, int right);
using namespace std;
int main()
{
  int i;
  char str[][10] = {"this","is","a","project"};
  cout<<"Before Sort"<<"\n";
  for(i=0;i<4;i++)
  {
      cout<<str[i]<<" "<<"\n";
  }
  cout<<"\n\n";
  quickSortMain(str, 4);
  cout<<"After Sort"<<"\n";
  for(i=0; i<4; i++) {
     
     cout<<str[i]<<" "<<"\n";
  }
  return 0;
}
void quickSortMain(char items[][10], int count)
{
  quickSort(items, 0, count-1);
}

void quickSort(char items[][10], int left, int right)
{
  int i, j;
  char *x;
  char temp[10];

  i = left;
  j = right;
  x = items[(left+right)/2];

  do {
    while((strcmp(items[i],x) < 0) && (i < right)) {
       i++;
    }
    while((strcmp(items[j],x) > 0) && (j > left)) {
        j--;
    }
    if(i <= j) {
      strcpy(temp, items[i]);
      strcpy(items[i], items[j]);
      strcpy(items[j], temp);
      i++;
      j--;
   }
  } while(i <= j);

  if(left < j) {
     quickSort(items, left, j);
  }
  if(i < right) {
     quickSort(items, i, right);
  }
}

File: PartitionStr.java
//Array Partition for Strings in java
import java.io.*;
import java.util.*;
package quicksortstring;

public class QuickSortString{

	private static String []a;
	public static void main(String[] args) {
	                
		a = new String[]{"Z","A","V","L"};
		System.out.println("Original String:");
		
		printArray();
                System.out.println("");
		
		sort();
		
		System.out.println("Sorted String:");
		
		//prints the sorted array
		printArray();
		
	}
 
	public static void sort(){
		int left = 0;
		int right = a.length-1;
			
		quickSort(left, right);
	}
	
	private static void quickSort(int left,int right){
		
		if(left >= right)
			return;
		
		String pivot = getMedian(left, right);
		int partition = partition(left, right, pivot);
		quickSort(0, partition-1);
		quickSort(partition+1, right);
	}

	private static int partition(int left,int right,String pivot){
		int leftCursor = left-1;
		int rightCursor = right;
		while(leftCursor < rightCursor){
	while(((Comparable<String>)a[++leftCursor]).compareTo(pivot) < 0);
        while(rightCursor > 0 && ((Comparable<String>)a[--rightCursor]).compareTo(pivot) > 0);
			if(leftCursor >= rightCursor){
				break;
			}else{
				swap(leftCursor, rightCursor);
			}
		}
		swap(leftCursor, right);
		return leftCursor;
	}
	
	public static String getMedian(int left,int right){
		int center = (left+right)/2;
		
		if(((Comparable<String>)a[left]).compareTo(a[center]) > 0)
			swap(left,center);
		
		if(((Comparable<String>)a[left]).compareTo(a[right]) > 0)
			swap(left, right);
		
		if(((Comparable<String>)a[center]).compareTo(a[right]) > 0)
			swap(center, right);
		
		swap(center, right);
		return a[right];
	}
`		public static void swap(int left,int right){
		String temp = a[left];
		a[left] = a[right];
		a[right] = temp;
	}
	
	public static void printArray(){
		for(String i : a){
			System.out.print(i+" ");
		}
